Context Service in LWC (#6)
salesforce/lwc-rfcs: text/0000-context-service.md

_Note: No html file is required, unless the author want to customize the UI of the context provider._

This class also provides a static accessor called `Provider`, which is intended to be used for the consumer to wire to the provided context, e.g.:

```js
import { LightningElement } from 'lwc';
import ThemeContext from 'x/themeContext';

// Consumer
export default class XBar extends LightningElement {
    @wire(ThemeContext.Provider) x;
}
```

### Providing a context value

Setting a context value must be an operation involving a custom element. Any element in the  light DOM subtree of that element or their descendent, can consume the value set for a context.

To provide a new context value, you must use the declarative form via an HTML template. Any structure that involves `<x-theme-context>` and `<x-bar>` components can share the `theme` value, e.g.: 

```html
<template>
    <x-theme-context theme="dark">
        <x-bar></x-bar>
    </x-theme-context>
</template>
```

It means that `<x-bar>` or any of its descendants can wire to `ThemeContext.Provider` to obtain the `theme` value.

## Implementation Details

> In order to implement `LightningContext`, it requires a reform on the the wire protocol. This reforms has two folds:

Reviewer: ```suggestion
In order to implement `LightningContext`, it requires a reform on the wire protocol. This reform has two folds:
```

> 1. We need to introduce a new event that can be dispatched by a wire adapter to create a side channel between the `eventTarget` with a context provider installed on an element in the the composed path. The new event is implemented via a new constructor called `LinkContextEvent`, and it expects a unique `token`, and a `callback` to finalize the linking process. E.g.:

Reviewer: ```suggestion
1. We need to introduce a new event that can be dispatched by a wire adapter to create a side channel between the `eventTarget` with a context provider installed on an element in the composed path. The new event is implemented via a new constructor called `LinkContextEvent`, and it expects a unique `token`, and a `callback` to finalize the linking process. E.g.:
```

```js
import { register, ValueChangedEvent, LinkContextEvent } from 'wire-service';

const adapter = Symbol('ContextProvider');
register(adapter, (eventTarget) => {
    let unsubscribeCallback;

    function callback(data, unsubscribe) {
        // provider calls this function at any given time
        eventTarget.dispatchEvent(new ValueChangedEvent(data));
        unsubscribeCallback = unsubscribe;
    }

    eventTarget.addEventListener('connect', () => {
        // attempt to link to a context provider identified by token

>         const event = new LinkContextEvent(token, callback);

Reviewer: What is `token`? It's not defined in this example.

Author: Excellent question. Maybe I should have a section under implementation details for just that.

`token` is the dom event name used to link to the parent element (maybe `token` is the wrong term here). I didn't want to call it `domEventName` because it might be confusing. The point of this event name is that it "should" be tokenized somehow... generated once during the definition of the adapter. In the case of `LightningContext` element, it uses that to create the adapter and to listen for it on the DOM, and it is created per adapter, as a unique event name.

Reviewer: So it's a [DOMString](https://developer.mozilla.org/en-US/docs/Web/API/DOMString) whose value is generated automatically by the system.

Author: yes, it is a DOMString but whose value must be provided by the author of the adapter, and we don't do much with it... other than dispatching a dom event with a callback whenever the adapter emits a `LinkContextEvent` event.

A third goal is to support the provision of context values on any element, whether it is LWC or not.

# Prior Art

* React Context: https://reactjs.org/docs/context.html

# Detailed design

Important Notes:

* It is important for the consumer to be able to consume more than one context value
from different context providers if needed.
* It is important to guarantee that the provisioned object is reactive, so the UI can
rely on it directly.

## Proposal: Context Consumer

This proposal provides a high-level API (an abstraction layer) for authors to consume a context value at any level in the DOM flatten tree. This must be a future-proof API, and must be easy to use and easy to maintain. Here is an example:

```js
import { LightningElement } from 'lwc';
import { XFooContext } from 'x/fooContext';
import { XBarContext } from 'x/barContext';

export default class MyComponent extends LightningElement {
    @wire(XFooContext) a; // context object provided by an element up in the flatten tree
    @wire(XBarContext) b; // context object provided by element up in the flatten tree
}
```

Pros:
* it could be statically verified when wired to a field.

> * it works nicely with inheritance since it relies on field declarations.

Reviewer 2: ```suggestion
* it works well with inheritance since it relies on field declarations.
```

> * it is extremely familiar since it relies on the `wire` decorator.

Reviewer 2: ```suggestion
* it is familiar since it relies on the `wire` decorator.
```

Cons:
* the wire decorator is often seen as more complicated, and often requires authors to be defensive in case the value is never resolved (which is not the case of for context).
* it requires changes in the wire protocol to expand its capabilities to support wiring based on flatten DOM tree. (more on this below)

_Note: This is very similar to how React does it, but extending it to access multiple contexts._

_Note: since `@wire` supports wiring to a method, it will work the same way for context._

## Proposal: Context Provider

> This proposal provides a low-level API for authors to provide a context value at any level in the flattened DOM tree that can be consume by any node in the subtree. This API is subject to changes as we understand more and more about how context is used.

Reviewer 2: ```suggestion
This proposal provides a low-level API for authors to provide a context value at any level in the flattened DOM tree that can be consumed by any node in the subtree. This API is subject to change as we understand more and more about how context is used.
```

### Defining a new Context

> The definition of a new context is equivalent to create a new wire adapter. It has an identity, and it has some internal implementation details that are necessary to work as the first argument of the `@wire` decorator.

Reviewer 2: ```suggestion
The definition of a new context is equivalent to creating a new wire adapter. It has an identity, and it has some internal implementation details that are necessary to work as the first argument of the `@wire` decorator.
```

## Other examples

### How to implement a React-like context HOC in LWC?

```js
import { LightningElement, createContextExperimental, setContextExperimental } from 'lwc';

// creating a new context definition
const Context = createContextExperimental({
    x: 0,
    y: 0,
});

export default class XFooContext extends LightningElement {
    _value = {};

    @api
    get theme() {
        return this._value;
    }
    set theme(v) {
        // you can validate `v` here if you want
        this._value = v;
        // setting a new context value 
        this.setNewContextValue();
    }

    connectedCallback() {
        // setting the initial value of the context
        this.setNewContextValue();
    }
    setNewContextValue() {
        setContextExperimental(this, Context, this._value);
    }
    static Consumer = Context;
}
```

The way a provider is used in a template is very straight forward:

```html
<template>
    <x-foo-context theme="dark">
        <c-consumer></c-consumer> 
    </x-foo-context>
    <x-foo-context value="lite">
        <c-other-consumer></c-other-consumer> 
    </x-foo-context>    
</template>
```

> Where `<c-consumer>` and `<c-other-consumer>` or any of their descendent will be able to consume the context value by using the `@wire(XFooContext.Consumer)` decorator, e.g.:

Reviewer 2: ```suggestion
Where `<c-consumer>` and `<c-other-consumer>` or any of their descendants will be able to consume the context value by using the `@wire(XFooContext.Consumer)` decorator, e.g.:
```

```js
import { LightningElement } from 'lwc';
import XFooContext from 'x/fooContext';

export default class MyComponent extends LightningElement {

>     @wire(XFooContext.Consumer) foo; // context object provided by <x-foo-context> element

Author: That is the whole point, we are not opinionated on how you share the adapter, it is up to you to decide how. In this case, `XFooContext` is a constructor for `<x-foo-context>` element that provisions the theme value, and `.Consumer` is the adapter that will be wired to the nearest `<x-foo-context>` element.

>     @wire(XFooContext.Consumer) foo; // context object provided by <x-foo-context> element

Author: also, this is to demonstrate that with this infrastructure you can implement the exact same thing you do in react.

}
```

# Other Alternatives

> * The two new APIs can be imported from `@lwc/context` or `@lwc/wire-service` rather than `lwc` directly. This might help to eventually allow others to use contexts with vanilla web components. It will also make it easier for Lightning Platform to restrict who has access to these APIs in platform.

Reviewer 2: ```suggestion
* The two new APIs can be imported from `@lwc/context` or `@lwc/wire-service` rather than `lwc` directly. This might help to eventually allow others to use contexts with vanilla web components. It will also make it easier for Lightning Platform to restrict who has access to these APIs on platform.
```

* Make the new APIs a lot more generic, instead of context specific, where you might want to use them to provide contextual wiring per adapters. The fear here is that we will give them a very powerful experimental API.

# Adoption strategy

This is a brand new feature, we just need to document it.

# How we teach this

* The consumer aspect of it is very straight forward considering that we have plenty of documentation about the `@wire` decorator.

> # Unresolved questions

Reviewer 3: A couple more:
- Is the value always synchronous provisioned during the `wiring` stage of the component lifecycle?  
- Is there a formal error state that the wire can provision? Will it participate in the component error boundary?

Author: added

* Should the context value be validated?
* Should the context value be type any? or must be an object?

> * Should the context value be serializable to support locker?

Reviewer 3: On Lightning Platform we've agreed to only provision inert data via a wire.
